### DISCUSSION TITLE
help me explain  esmodules in nodejs

### DISCUSSION THREAD

**Discussion by WingDust** [nodejs/help#4946] [Q&A] [answered]
What is esmodules in nodejs ?
That is commonjs another law? 
What is the difference between esmodules and commonjs in nodejs? 

I read doc and still don't understand

**Comment by F3n67u** [comment]
You can take a look at this tutorial : https://blog.logrocket.com/commonjs-vs-es-modules-node-js/

**Reply by WingDust** [reply]
it don't answer the question , it 's more like about use

**Comment by VoltrexKeyva** [comment, marked as answer]
CommonJS (CJS) was implemented as a way to modularize JavaScript code when JavaScript didn't officially have a way to do so, but now JavaScript has an official way to modularize code through ES (ECMAScript) modules (also known as ESM).

CJS isn't the only way to modularize code when JavaScript didn't have an official way to do so, there is also [SystemJS](https://github.com/systemjs/systemjs),  [RequireJS](https://requirejs.org/), AMD (explained in RequireJS), etc etc. Although Node.js mainly uses CommonJS instead of the other ones.

The main differences are as follows:
1. CJS module evaluation is not asynchronous, and does not support features such as [Top-Level Await (TLA)](https://v8.dev/features/top-level-await).
2. Modules can be imported in CJS using both the global [`require()`](https://nodejs.org/api/modules.html#requireid) function and [dynamic imports (`import()`)](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/import#dynamic_imports) while modules can be imported in ES modules using the `import` statements (`import ... from ...`) and dynamic imports. (the `require()` function does not exist by default on the global scope in ESM, but can be made using the [`createRequire()`](https://nodejs.org/api/module.html#modulecreaterequirefilename) function from the [`node:module` built-in module](https://nodejs.org/api/module.html))
3. Code can be exported in CJS by assigning properties/methods to the `module.exports` property (or completely re-assigning the `exports` property to whatever), while code can be exported in ESM using the `export` keyword. (and the `module` and `exports` objects does not exist on the global scope in an ES module)
5. CJS's global `require()` function doesn't require the extension of a file when importing the said file while ESM's `import` statements and dynamic imports does (dynamic imports also exist in CJS), the reason ESM's `import` statements and dynamic imports requires the file extension is to make it less confusing not only for the module loaders to decide on which file to load, but also help developers who doesn't have an exact grasp of the code they're reading to understand from which file things are being imported from.
6. CJS's global `require()` function can import [C++ addons](https://nodejs.org/api/addons.html) while ESM's `import` statements or dynamic imports can't, although the `require()` function can be created in ESM as stated in point 2.
7. CJS's global `require()` function can directly import JSON modules, while ESM's `import` statements and dynamic imports require an [import assertion](https://v8.dev/features/import-assertions) to do so.
8. ESM's module cache is not exposed to the user-land so the users can't modify the cache at will for security reasons while CJS's module cache is exposed and can be modified in any way desired (through the [`require.cache` property](https://nodejs.org/api/modules.html#requirecache)).
9. CJS's global `require()` function can't import ES modules (although you can import ES modules in CJS using dynamic imports).

There are more differences but these are the main differences that are focused upon.

Overall it's recommended to use JavaScript's official way to modularize code (ESM) for increased stability and security, and the extended features it provides, if you're heading for backwards compatibility then you should still use CJS rather than ESM, if that's not your main focus, then ESM is the better choice.

**Comment by ShrikantaMazumder** [comment]
Let's see this one. 
[Here](https://michaelcurrin.github.io/dev-cheatsheets/cheatsheets/javascript/general/modules/es-modules.html#:~:text=As%20of%20ES6%20(ES2015)%2C,exports%20and%20require%20.)
